### PR TITLE
JAMES-3382 Email/query position and limit

### PR DESCRIPTION
**Pull request by LanKhuat**
All work has already been done by @remk. Just added some test to meet ticket requirements.

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 1]
```diff
@@ -1912,6 +1912,136 @@ trait EmailQueryMethodContract {
...
+           |                "canCalculateChanges": false,
+           |                "position": 2,
+           |                "ids": []
+           |            },
+           |            "c1"
+           |        ]]
+           |}""".stripMargin)
+    }
+  }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: Can you test 5 messages, position 2, limit 2 ?

Comment (LanKhuat) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: Should still be empty? Is there any particular goal you want to achieve with this setup?

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: Given 1 2 3 4 5 message, query with limit(2) position(2) and return `3 4`

Comment (remk) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: Yes i agree with @chibenwa. You should change the name of the test.

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: I neant "add a test"

Comment (LanKhuat) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: Hmm I wonder what should be applied first in the query? Position > Limit or Limit > Position? If it is the former then we need some change in the current logic.

Comment (LanKhuat) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: Currently it is Limit first then Position. So with above setup 1 2 3 4 5 will return (5 4) 3 2 1 then applying Position we will have an empty array.

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:2044]: position first then limit :-)

In previous draft code:

```
results.skip(position).limit(limit)
```

**Comment by LanKhuat** [comment]
Not sure if this hackish approach is acceptable or we want to make changes to the StoreMailboxManager to handle position. Suggestion?

**Comment by chibenwa** [comment]
test this please

**Review by chibenwa — COMMENTED** [review]
Okayish to me

test this please

**Comment by chibenwa** [comment]
test this please

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Comment by LanKhuat** [comment]
test this please

**Review by remk — APPROVED** [review]

**Comment by Arsnael** [comment]
Need to rebase

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**